File: packages/shared/src/types/goal.ts
import type { DateString, DateTimeString, ID } from "./user";

export type GoalStatus = "draft" | "active" | "completed" | "paused" | "archived";

/** User-defined goal — no fixed duration; pace is personal */
export interface Goal {
  id: ID;
  userId: ID;
  futureSelfId?: ID;
  title: string;
  description: string;
  /** Optional milestone the user sets for themselves */
  targetDate?: DateString;
  status: GoalStatus;
  createdAt: DateTimeString;
  updatedAt: DateTimeString;
}

export type PlanStatus = "draft" | "active" | "completed" | "superseded";

/** Personal plan or process step collection — flexible length, user-driven */
export interface Plan {
  id: ID;
  userId: ID;
  goalId?: ID;
  title: string;
  description?: string;
  status: PlanStatus;
  version: number;
  createdAt: DateTimeString;
  updatedAt: DateTimeString;
}

export type TaskStatus = "pending" | "in_progress" | "completed" | "skipped";

/** Single personalized micro-step for a given day */
export interface DailyTask {
  id: ID;
  userId: ID;
  planId: ID;
  date: DateString;
  title: string;
  description: string;
  status: TaskStatus;
  completedAt?: DateTimeString;
  /** Skipping does not penalize alignment score */
  skippedReason?: string;
  createdAt: DateTimeString;
  updatedAt: DateTimeString;
}

export type WeeklyReviewStatus = "pending" | "generating" | "ready" | "applied";

/** AI-generated weekly evaluation that may produce the next Plan version */
export interface WeeklyReview {
  id: ID;
  userId: ID;
  planId: ID;
  weekStartDate: DateString;
  summary: string;
  adaptations: string[];
  nextPlanVersion?: number;
  status: WeeklyReviewStatus;
  createdAt: DateTimeString;
  updatedAt: DateTimeString;
}

File: apps/api/internal/domain/goal/model/goal.go
package model

import (
	"time"

	"github.com/google/uuid"
)

type GoalStatus string

const (
	GoalDraft     GoalStatus = "draft"
	GoalActive    GoalStatus = "active"
	GoalCompleted GoalStatus = "completed"
	GoalPaused    GoalStatus = "paused"
	GoalArchived  GoalStatus = "archived"
)

type Goal struct {
	ID           uuid.UUID
	UserID       uuid.UUID
	FutureSelfID *uuid.UUID
	Title        string
	Description  string
	TargetDate   *time.Time
	Status       GoalStatus
	CreatedAt    time.Time
	UpdatedAt    time.Time
}

type PlanStatus string

const (
	PlanDraft      PlanStatus = "draft"
	PlanActive     PlanStatus = "active"
	PlanCompleted  PlanStatus = "completed"
	PlanSuperseded PlanStatus = "superseded"
)

type Plan struct {
	ID          uuid.UUID
	UserID      uuid.UUID
	GoalID      *uuid.UUID
	Title       string
	Description *string
	Status      PlanStatus
	Version     int
	Steps       []PlanStep
	CreatedAt   time.Time
	UpdatedAt   time.Time
}

type PlanStep struct {
	ID          uuid.UUID
	PlanID      uuid.UUID
	DayOffset   int
	Title       string
	Description string
	SortOrder   int
}

type TaskStatus string

const (
	TaskPending    TaskStatus = "pending"
	TaskInProgress TaskStatus = "in_progress"
	TaskCompleted  TaskStatus = "completed"
	TaskSkipped    TaskStatus = "skipped"
)

type DailyTask struct {
	ID            uuid.UUID
	UserID        uuid.UUID
	PlanID        uuid.UUID
	Date          time.Time
	Title         string
	Description   string
	Status        TaskStatus
	CompletedAt   *time.Time
	SkippedReason *string
	CreatedAt     time.Time
	UpdatedAt     time.Time
}
